ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Ledger/Tables.hs

```diff
@@ -0,0 +1,388 @@
+    => l any
+    -> LedgerTables l mk
+    -> l mk
+  withLedgerTables st _ = convertMapKind st
+
+instance ( Ord (Key l)
+         , Eq (Value l)
+         , Show (Key l)
+         , Show (Value l)
+         , NoThunks (Key l)
+         , NoThunks (Value l)
+         ) => HasLedgerTables (LedgerTables l) where
+  projectLedgerTables = castLedgerTables
+  withLedgerTables _ = castLedgerTables
+
+-- | Convenience class, useful for partially applying the composition of
+-- 'HasLedgerTables' and 'Ticked1'.
+type HasTickedLedgerTables :: LedgerStateKind -> Constraint
+class HasLedgerTables (Ticked1 l) => HasTickedLedgerTables l where
+instance HasLedgerTables (Ticked1 l) => HasTickedLedgerTables l
+
+-- | LedgerTables are projections of data from a LedgerState and as such they
+-- can be injected back into a LedgerState. This is necessary because the Ledger
+-- rules are unaware of UTxO-HD changes. Thus, by stowing the ledger tables, we are
+-- able to provide a Ledger State with a restricted UTxO set that is enough to
+-- execute the Ledger rules.
+--
+-- In particular, HardForkBlock LedgerStates are never given diretly to the
+-- ledger but rather unwrapped and then it is the inner ledger state the one we
+-- give to the ledger. This means that all the single era blocks must be an
+-- instance of this class, but HardForkBlocks might avoid doing so.
+type CanStowLedgerTables :: LedgerStateKind -> Constraint
+class CanStowLedgerTables l where
+
+  stowLedgerTables     :: l ValuesMK -> l EmptyMK
+  default stowLedgerTables ::
+       (LedgerTablesAreTrivial l)
+    => l ValuesMK
+    -> l EmptyMK
+  stowLedgerTables = convertMapKind
+
+  unstowLedgerTables   :: l EmptyMK  -> l ValuesMK
+  default unstowLedgerTables ::
+       (LedgerTablesAreTrivial l)
+    => l EmptyMK
+    -> l ValuesMK
+  unstowLedgerTables = convertMapKind
+
+{-------------------------------------------------------------------------------
+  Serialization Codecs
+-------------------------------------------------------------------------------}
+
+-- | This class provides a 'CodecMK' that can be used to encode/decode keys and
+-- values on @'LedgerTables' l mk@
+--
+-- TODO: can this be removed in favour of EncodeDisk and DecodeDisk?
+type CanSerializeLedgerTables :: LedgerStateKind -> Constraint
+class CanSerializeLedgerTables l where
+  codecLedgerTables :: LedgerTables l CodecMK
+  default codecLedgerTables ::
+       ( FromCBOR (Key l), FromCBOR (Value l)
+       , ToCBOR   (Key l), ToCBOR   (Value l)
+       )
+    => LedgerTables l CodecMK
+  codecLedgerTables = LedgerTables $ CodecMK toCBOR toCBOR fromCBOR fromCBOR
+
+-- | Default encoder of @'LedgerTables' l ''ValuesMK'@ to be used by the
+-- in-memory backing store.
+valuesMKEncoder ::
+     ( HasLedgerTables l
+     , CanSerializeLedgerTables l
+     )
+  => LedgerTables l ValuesMK
+  -> CBOR.Encoding
+valuesMKEncoder tables =
+       CBOR.encodeListLen (ltcollapse $ ltmap (K2 . const 1) tables)
```

Comment:
`ltcollapse` is the ledger tables equivalent of `collapse_NS`. If ledger tables were to become a product type, then the type of `ltcollapse` would probably change to resemble `collapse_NP`. The `HCollapse` class is in general used for "collapsing heterogenous structures into homogeneous ones".

```
collapse_NS :: NS (K a) xs -> a
collapse_NP :: NP (K a) xs -> [a]
```

I think that the use of `ltcollapse` here is intendend for future proofing this code. It's at least more future proof than simply hard-coding `1` as the list length. If want to use a fold, then we should create an `ltfoldMap` equivalent to https://hackage.haskell.org/package/sop-core-0.5.0.2/docs/Data-SOP.html#v:hcfoldMap. Alternatively, if `ltcollapse` were to return a list in the feature, then we could `foldMap` over the list afterwards

Reply:
I had initially decided it was best to replace uses of `ltcollapse` by some new `ltfoldMap` function, but after some thinking it's best to instead use `ltcollapse` as is, but removing the use of monoids there, which currently has no effect.

The reason I think this is the right call is because ledger tables are currently a single-constructor newtype, and they will be for at least a while. We do not know what ledger tables will look like when we store more parts of the ledger state, so let's cross that bridge when we get there.